# PART - 1

### Importing the necessary libraries

In [23]:
import numpy as np # library to handle data in a vectorized manner
import pandas as pd # library for data analsysis

import json # library to handle JSON files

!conda install -c conda-forge geopy --yes 
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

!conda install -c conda-forge folium=0.5.0 --yes 
import folium # map rendering library

print('Libraries imported.')

Solving environment: done

## Package Plan ##

  environment location: /opt/conda/envs/Python36

  added / updated specs: 
    - geopy


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    openssl-1.1.1f             |       h516909a_0         2.1 MB  conda-forge
    python_abi-3.6             |          1_cp36m           4 KB  conda-forge
    geopy-1.21.0               |             py_0          58 KB  conda-forge
    ca-certificates-2020.4.5.1 |       hecc5488_0         146 KB  conda-forge
    certifi-2020.4.5.1         |   py36h9f0ad1d_0         151 KB  conda-forge
    geographiclib-1.50         |             py_0          34 KB  conda-forge
    ------------------------------------------------------------
                                           Total:         2.5 MB

The following NEW packages will be INSTALLED:

    geographiclib:   1.50-py_0         conda-forge
    geopy:           1

In [70]:
from pandas.io.html import read_html

### Creating a DataFrame from the table given at the webpage

In [74]:
page = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"

In [75]:
wikitables = read_html(page, attrs={"class": "wikitable"})

In [76]:
df = wikitables[0]

In [77]:
df.head()

Postal code           Borough                Neighborhood
0         M1A      Not assigned                         NaN
1         M2A      Not assigned                         NaN
2         M3A        North York                   Parkwoods
3         M4A        North York            Victoria Village
4         M5A  Downtown Toronto  Regent Park / Harbourfront

In [78]:
df.columns = ["PostalCode", "Borough", "Neighborhood"]

In [79]:
df.head()

PostalCode           Borough                Neighborhood
0        M1A      Not assigned                         NaN
1        M2A      Not assigned                         NaN
2        M3A        North York                   Parkwoods
3        M4A        North York            Victoria Village
4        M5A  Downtown Toronto  Regent Park / Harbourfront

### Removing the rows from Borough column that are not assigned

In [80]:
df.drop(df.index[df["Borough"] == "Not assigned"], inplace = True)

In [81]:
df.head()

PostalCode           Borough                                  Neighborhood
2        M3A        North York                                     Parkwoods
3        M4A        North York                              Victoria Village
4        M5A  Downtown Toronto                    Regent Park / Harbourfront
5        M6A        North York             Lawrence Manor / Lawrence Heights
6        M7A  Downtown Toronto  Queen's Park / Ontario Provincial Government

In [82]:
#fixing the format of Neighborhood column
df["Neighborhood"] = df['Neighborhood'].apply(lambda x : x.replace('/', ','))
df.head(10)

PostalCode           Borough                                  Neighborhood
2         M3A        North York                                     Parkwoods
3         M4A        North York                              Victoria Village
4         M5A  Downtown Toronto                    Regent Park , Harbourfront
5         M6A        North York             Lawrence Manor , Lawrence Heights
6         M7A  Downtown Toronto  Queen's Park , Ontario Provincial Government
8         M9A         Etobicoke                              Islington Avenue
9         M1B       Scarborough                               Malvern , Rouge
11        M3B        North York                                     Don Mills
12        M4B         East York              Parkview Hill , Woodbine Gardens
13        M5B  Downtown Toronto                      Garden District, Ryerson

### Combining Neighbourhoods based on similar Postcode and Borough

In [83]:
df = df.groupby(['PostalCode', 'Borough'])['Neighborhood'].apply(','.join).reset_index()
df.columns = ['PostalCode','Borough','Neighborhood']
df.head(10)

PostalCode      Borough                                       Neighborhood
0        M1B  Scarborough                                    Malvern , Rouge
1        M1C  Scarborough           Rouge Hill , Port Union , Highland Creek
2        M1E  Scarborough                Guildwood , Morningside , West Hill
3        M1G  Scarborough                                             Woburn
4        M1H  Scarborough                                          Cedarbrae
5        M1J  Scarborough                                Scarborough Village
6        M1K  Scarborough      Kennedy Park , Ionview , East Birchmount Park
7        M1L  Scarborough                  Golden Mile , Clairlea , Oakridge
8        M1M  Scarborough  Cliffside , Cliffcrest , Scarborough Village West
9        M1N  Scarborough                       Birch Cliff , Cliffside West

### Assigning Borough values to the Neignbourhood where vlaue is "Not assigned" or "NaN"

In [84]:
df.loc[df['Neighborhood'] == 'Not assigned', 'Neighborhood'] = df['Borough']
df['Neighborhood'].fillna(df["Borough"], inplace = True)

In [85]:
df.head()

PostalCode      Borough                              Neighborhood
0        M1B  Scarborough                           Malvern , Rouge
1        M1C  Scarborough  Rouge Hill , Port Union , Highland Creek
2        M1E  Scarborough       Guildwood , Morningside , West Hill
3        M1G  Scarborough                                    Woburn
4        M1H  Scarborough                                 Cedarbrae

In [86]:
df.shape

(103, 3)

# PART - 2

### Using the csv given in the link

In [87]:
link = "http://cocl.us/Geospatial_data"
df1 = pd.read_csv(link)

df1.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [89]:
#checking the shape 
df1.shape

(103, 3)

Both the dataframes have 103 rows and 3 columns, so we can merge them 


In [90]:
df1.columns = ["PostalCode", "Latitude", "Longitude"]

In [91]:
df1.head()

PostalCode   Latitude  Longitude
0        M1B  43.806686 -79.194353
1        M1C  43.784535 -79.160497
2        M1E  43.763573 -79.188711
3        M1G  43.770992 -79.216917
4        M1H  43.773136 -79.239476

### Merging the two dataframes based on their PostalCode

In [92]:
df_new = pd.merge(df, df1, on='PostalCode')
df_new.head()

PostalCode      Borough                              Neighborhood  \
0        M1B  Scarborough                           Malvern , Rouge   
1        M1C  Scarborough  Rouge Hill , Port Union , Highland Creek   
2        M1E  Scarborough       Guildwood , Morningside , West Hill   
3        M1G  Scarborough                                    Woburn   
4        M1H  Scarborough                                 Cedarbrae   

    Latitude  Longitude  
0  43.806686 -79.194353  
1  43.784535 -79.160497  
2  43.763573 -79.188711  
3  43.770992 -79.216917  
4  43.773136 -79.239476

In [93]:
neighborhoods = df_new

# PART-3

### finding how many boroughs and neighborhood in the data

In [94]:
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(neighborhoods['Borough'].unique()),
        neighborhoods.shape[0]
    )
)

The dataframe has 10 boroughs and 103 neighborhoods.


### Use geopy library to get the latitude and longitude values of Toronto, Canada

In [95]:
address = 'Toronto, Canada'

geolocator = Nominatim(user_agent="can_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


### create a map of toronto to visualize the neighborhood

In [96]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(neighborhoods['Latitude'], neighborhoods['Longitude'], neighborhoods['Borough'], neighborhoods['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

### defining fouraquare API credentials and version


In [97]:
LIMIT = 100

CLIENT_ID = 'GXB3HZYAZFDTWD540G3OLDGPM54YHJKEDGHS5NM5W2Y2HMO4' # your Foursquare ID
CLIENT_SECRET = 'ZXQAVEYIWIHJNDMKEVGCG0XJ2ZGWO52LUYGJU2DWPQAM5X4N' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: GXB3HZYAZFDTWD540G3OLDGPM54YHJKEDGHS5NM5W2Y2HMO4
CLIENT_SECRET:ZXQAVEYIWIHJNDMKEVGCG0XJ2ZGWO52LUYGJU2DWPQAM5X4N


## Exploring neighborhoods in Toronto

### Let's create a function to repeat the same process to all the neighborhoods in Toronto

In [98]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

### Now write the code to run the above function on each neighborhood and create a new dataframe called toronto_venues

In [99]:
toronto_venues = getNearbyVenues(names=neighborhoods['Neighborhood'],
                                   latitudes=neighborhoods['Latitude'],
                                   longitudes=neighborhoods['Longitude'],
                                  )

Malvern , Rouge
Rouge Hill , Port Union , Highland Creek
Guildwood , Morningside , West Hill
Woburn
Cedarbrae
Scarborough Village
Kennedy Park , Ionview , East Birchmount Park
Golden Mile , Clairlea , Oakridge
Cliffside , Cliffcrest , Scarborough Village West
Birch Cliff , Cliffside West
Dorset Park , Wexford Heights , Scarborough Town Centre
Wexford , Maryvale
Agincourt
Clarks Corners , Tam O'Shanter , Sullivan
Milliken , Agincourt North , Steeles East , L'Amoreaux East
Steeles West , L'Amoreaux West
Upper Rouge
Hillcrest Village
Fairview , Henry Farm , Oriole
Bayview Village
York Mills , Silver Hills
Willowdale , Newtonbrook
Willowdale
York Mills West
Willowdale
Parkwoods
Don Mills
Don Mills
Bathurst Manor , Wilson Heights , Downsview North
Northwood Park , York University
Downsview
Downsview
Downsview
Downsview
Victoria Village
Parkview Hill , Woodbine Gardens
Woodbine Heights
The Beaches
Leaside
Thorncliffe Park
East Toronto
The Danforth West , Riverdale
India Bazaar , The Beaches 

In [100]:
# new dataset
toronto_venues.head()

Neighborhood  Neighborhood Latitude  \
0                           Malvern , Rouge              43.806686   
1  Rouge Hill , Port Union , Highland Creek              43.784535   
2  Rouge Hill , Port Union , Highland Creek              43.784535   
3       Guildwood , Morningside , West Hill              43.763573   
4       Guildwood , Morningside , West Hill              43.763573   

   Neighborhood Longitude                  Venue  Venue Latitude  \
0              -79.194353                Wendy’s       43.807448   
1              -79.160497      RIGHT WAY TO GOLF       43.785177   
2              -79.160497  Royal Canadian Legion       43.782533   
3              -79.188711      G & G Electronics       43.765309   
4              -79.188711       Big Bite Burrito       43.766299   

   Venue Longitude        Venue Category  
0       -79.199056  Fast Food Restaurant  
1       -79.161108           Golf Course  
2       -79.163085                   Bar  
3       -79.191537     Electronics Store  
4       -79.190720    Mexican Restaurant

In [101]:
toronto_venues.shape

(2125, 7)

### checking how many venues were returned for each neighborhood

In [102]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Agincourt                                                               4   
Alderwood , Long Branch                                                10   
Bathurst Manor , Wilson Heights , Downsview North                      19   
Bayview Village                                                         4   
Bedford Park , Lawrence Manor East                                     26   
Berczy Park                                                            56   
Birch Cliff , Cliffside West                                            4   
Brockton , Parkdale Village , Exhibition Place                         23   
Business reply mail Processing CentrE                                  14   
CN Tower , King and Spadina , Railway Lands , H...                     18   
Caledonia-Fairbanks                                                     4   
Canada Post Gateway Processing Centre                                  13   
Cedarbrae                                                               8   
Central Bay Street                                                     60   
Christie                                                               18   
Church and Wellesley                                                   73   
Clarks Corners , Tam O'Shanter , Sullivan                              15   
Cliffside , Cliffcrest , Scarborough Village West                       2   
Commerce Court , Victoria Hotel                                       100   
Davisville                                                             34   
Davisville North                                                        7   
Del Ray , Mount Dennis , Keelsdale and Silverthorn                      4   
Don Mills                                                              27   
Dorset Park , Wexford Heights , Scarborough Tow...                      5   
Downsview                                                              13   
Dufferin , Dovercourt Village                                          18   
East Toronto                                                            3   
Eringate , Bloordale Gardens , Old Burnhamthorp...                      8   
Fairview , Henry Farm , Oriole                                         67   
First Canadian Place , Underground city                               100   
...                                                                   ...   
Richmond , Adelaide , King                                             94   
Rosedale                                                                4   
Roselawn                                                                2   
Rouge Hill , Port Union , Highland Creek                                2   
Runnymede , Swansea                                                    36   
Runnymede , The Junction North                                          3   
Scarborough Village                                                     1   
South Steeles , Silverstone , Humbergate , Jame...                     11   
St. James Town                                                         80   
St. James Town , Cabbagetown                                           43   
Steeles West , L'Amoreaux West                                         15   
Stn A PO Boxes                                                         95   
Studio District                                                        40   
Summerhill West , Rathnelly , South Hill , Fore...                     16   
The Annex , North Midtown , Yorkville                                  23   
The Beaches                                                             6   
The Danforth West , Riverdale                                          43   
The Kingsway , Montgomery Road , Old Mill North                         3   
Thorncliffe Park                                                       20   
Toronto Dominion Centre , Design Exchange       

### checking how many unique categories

In [103]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 267 uniques categories.


## Analyze each neighborhood

In [104]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Yoga Studio  Accessories Store  Airport  Airport Food Court  Airport Gate  \
0            0                  0        0                   0             0   
1            0                  0        0                   0             0   
2            0                  0        0                   0             0   
3            0                  0        0                   0             0   
4            0                  0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  ...  Trail  Train Station  Vegetarian / Vegan Restaurant  \
0             0  ...      0              0                              0   
1             0  ...      0              0                              0   
2             0  ...      0              0                              0   
3             0  ...      0              0                              0   
4             0  ...      0              0                              0   

   Video Game Store  Video Store  Vietnamese Restaurant  Warehouse Store  \
0                 0            0                      0                0   
1                 0            0                      0                0   
2                 0            0                      0                0   
3                 0            0                      0                0   
4                 0            0                      0                0   

   Wine Bar  Wings Joint  Women's Store  
0         0            0              0  
1         0            0              0  
2         0            0              0  
3         0            0              0  
4         0            0              0  

[5 rows x 267 columns]

In [105]:
toronto_onehot.shape

(2125, 267)

### Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [106]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Yoga Studio  \
0                                           Agincourt     0.000000   
1                             Alderwood , Long Branch     0.000000   
2   Bathurst Manor , Wilson Heights , Downsview North     0.000000   
3                                     Bayview Village     0.000000   
4                  Bedford Park , Lawrence Manor East     0.000000   
5                                         Berczy Park     0.000000   
6                        Birch Cliff , Cliffside West     0.000000   
7      Brockton , Parkdale Village , Exhibition Place     0.000000   
8               Business reply mail Processing CentrE     0.000000   
9   CN Tower , King and Spadina , Railway Lands , ...     0.000000   
10                                Caledonia-Fairbanks     0.000000   
11              Canada Post Gateway Processing Centre     0.000000   
12                                          Cedarbrae     0.000000   
13                                 Central Bay Street     0.016667   
14                                           Christie     0.000000   
15                               Church and Wellesley     0.027397   
16          Clarks Corners , Tam O'Shanter , Sullivan     0.000000   
17  Cliffside , Cliffcrest , Scarborough Village West     0.000000   
18                    Commerce Court , Victoria Hotel     0.000000   
19                                         Davisville     0.000000   
20                                   Davisville North     0.000000   
21  Del Ray , Mount Dennis , Keelsdale and Silvert...     0.000000   
22                                          Don Mills     0.000000   
23  Dorset Park , Wexford Heights , Scarborough To...     0.000000   
24                                          Downsview     0.000000   
25                      Dufferin , Dovercourt Village     0.000000   
26                                       East Toronto     0.000000   
27  Eringate , Bloordale Gardens , Old Burnhamthor...     0.000000   
28                     Fairview , Henry Farm , Oriole     0.000000   
29            First Canadian Place , Underground city     0.000000   
..                                                ...          ...   
64                         Richmond , Adelaide , King     0.000000   
65                                           Rosedale     0.000000   
66                                           Roselawn     0.000000   
67           Rouge Hill , Port Union , Highland Creek     0.000000   
68                                Runnymede , Swansea     0.000000   
69                     Runnymede , The Junction North     0.000000   
70                                Scarborough Village     0.000000   
71  South Steeles , Silverstone , Humbergate , Jam...     0.000000   
72                                     St. James Town     0.000000   
73                       St. James Town , Cabbagetown     0.000000   
74                     Steeles West , L'Amoreaux West     0.000000   
75                                     Stn A PO Boxes     0.010526   
76                                    Studio District     0.025000   
77  Summerhill West , Rathnelly , South Hill , For...     0.000000   
78              The Annex , North Midtown , Yorkville     0.000000   
79                                        The Beaches     0.000000   
80                      The Danforth West , Riverdale     0.023256   
81    The Kingsway , Montgomery Road , Old Mill North     0.000000   
82                                   Thorncliffe Park     0.050000   
83          Toronto Dominion Centre , Design Exchange     0.000000   
84                    University of Toronto , Harbord     0.028571   
85                                   Victoria Village     0.000000   
86                                          Westmount     0.000000   
87                                             Weston     0.000000   
88                                 Wexford , Maryvale     0.000000   
89                                         Willowda

In [107]:
# confirm the new size
toronto_grouped.shape

(94, 267)

### Let's print each neighborhood along with the top 5 most common venues

In [108]:
num_top_venues = 5

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Agincourt----
                       venue  freq
0             Breakfast Spot  0.25
1  Latin American Restaurant  0.25
2               Skating Rink  0.25
3                     Lounge  0.25
4                Yoga Studio  0.00


----Alderwood , Long Branch----
            venue  freq
0     Pizza Place   0.2
1        Pharmacy   0.1
2            Pool   0.1
3  Sandwich Place   0.1
4             Pub   0.1


----Bathurst Manor , Wilson Heights , Downsview North----
            venue  freq
0            Bank  0.11
1     Coffee Shop  0.11
2        Pharmacy  0.05
3  Ice Cream Shop  0.05
4   Shopping Mall  0.05


----Bayview Village----
                 venue  freq
0                 Café  0.25
1                 Bank  0.25
2  Japanese Restaurant  0.25
3   Chinese Restaurant  0.25
4          Yoga Studio  0.00


----Bedford Park , Lawrence Manor East----
                venue  freq
0      Sandwich Place  0.08
1         Pizza Place  0.08
2          Restaurant  0.08
3    Sushi Restaurant  0.08
4  It

### Let's put that into a pandas dataframe

In [109]:
#function to sort the venues in descending order.
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

### Now let's create the new dataframe and display the top 10 venues for each neighborhood.

In [110]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood  \
0                                          Agincourt   
1                            Alderwood , Long Branch   
2  Bathurst Manor , Wilson Heights , Downsview North   
3                                    Bayview Village   
4                 Bedford Park , Lawrence Manor East   

       1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0  Latin American Restaurant          Skating Rink                Lounge   
1                Pizza Place                   Gym    Athletics & Sports   
2                       Bank           Coffee Shop              Pharmacy   
3                       Café                  Bank    Chinese Restaurant   
4           Sushi Restaurant        Sandwich Place    Italian Restaurant   

  4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0        Breakfast Spot         Women's Store            Donut Shop   
1              Pharmacy                  Pool                   Pub   
2           Supermarket         Deli / Bodega      Sushi Restaurant   
3   Japanese Restaurant    Dim Sum Restaurant        Discount Store   
4           Coffee Shop            Restaurant           Pizza Place   

  7th Most Common Venue      8th Most Common Venue 9th Most Common Venue  \
0        Discount Store        Distribution Center               Dog Run   
1        Sandwich Place               Skating Rink           Coffee Shop   
2                 Diner  Middle Eastern Restaurant            Restaurant   
3   Distribution Center                    Dog Run      Doner Restaurant   
4              Pharmacy               Liquor Store             Locksmith   

    10th Most Common Venue  
0         Doner Restaurant  
1      American Restaurant  
2      Fried Chicken Joint  
3               Donut Shop  
4  Comfort Food Restaurant

## Cluster Neighborhoods

In [111]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]

array([1, 3, 1, 3, 1, 1, 1, 1, 3, 1], dtype=int32)

In [112]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = neighborhoods

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head() # check the last columns!

PostalCode      Borough                              Neighborhood  \
0        M1B  Scarborough                           Malvern , Rouge   
1        M1C  Scarborough  Rouge Hill , Port Union , Highland Creek   
2        M1E  Scarborough       Guildwood , Morningside , West Hill   
3        M1G  Scarborough                                    Woburn   
4        M1H  Scarborough                                 Cedarbrae   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.806686 -79.194353             2.0  Fast Food Restaurant   
1  43.784535 -79.160497             1.0           Golf Course   
2  43.763573 -79.188711             3.0   Rental Car Location   
3  43.770992 -79.216917             1.0           Coffee Shop   
4  43.773136 -79.239476             3.0   Fried Chicken Joint   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0    Dim Sum Restaurant    Falafel Restaurant           Event Space   
1                   Bar         Women's Store             Drugstore   
2        Breakfast Spot    Mexican Restaurant     Electronics Store   
3     Korean Restaurant     Convenience Store         Women's Store   
4           Gas Station                  Bank      Hakka Restaurant   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0  Ethiopian Restaurant   Empanada Restaurant     Electronics Store   
1        Discount Store   Distribution Center               Dog Run   
2                  Bank          Intersection        Medical Center   
3             Drugstore        Discount Store   Distribution Center   
4  Caribbean Restaurant    Athletics & Sports       Thai Restaurant   

         8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0  Eastern European Restaurant   Dumpling Restaurant              Drugstore  
1             Doner Restaurant            Donut Shop    Dumpling Restaurant  
2             Doner Restaurant   Distribution Center                Dog Run  
3                      Dog Run      Doner Restaurant             Donut Shop  
4                       Bakery                 Diner         Discount Store

### Finally, let's visualize the resulting clusters

In [123]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        fill=True,
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## Examine Clusters

### cluster 1

In [124]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
5       Scarborough             0.0            Playground   
48  Central Toronto             0.0            Restaurant   
72       North York             0.0   Japanese Restaurant   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
5          Women's Store            Donut Shop    Dim Sum Restaurant   
48            Playground            Donut Shop    Dim Sum Restaurant   
72                Bakery                   Pub            Playground   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
5                  Diner        Discount Store   Distribution Center   
48                 Diner        Discount Store   Distribution Center   
72         Women's Store        Discount Store   Distribution Center   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
5                Dog Run      Doner Restaurant              Drugstore  
48               Dog Run      Doner Restaurant              Drugstore  
72               Dog Run      Doner Restaurant             Donut Shop

### cluster 2

In [125]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels      1st Most Common Venue  \
1        Scarborough             1.0                Golf Course   
3        Scarborough             1.0                Coffee Shop   
6        Scarborough             1.0             Discount Store   
7        Scarborough             1.0                     Bakery   
8        Scarborough             1.0        American Restaurant   
9        Scarborough             1.0            College Stadium   
11       Scarborough             1.0  Middle Eastern Restaurant   
12       Scarborough             1.0  Latin American Restaurant   
17        North York             1.0                       Pool   
18        North York             1.0             Clothing Store   
22        North York             1.0           Ramen Restaurant   
24        North York             1.0           Ramen Restaurant   
26        North York             1.0                        Gym   
27        North York             1.0                        Gym   
28        North York             1.0                       Bank   
29        North York             1.0       Caribbean Restaurant   
34        North York             1.0              Grocery Store   
37      East Toronto             1.0                        Pub   
38         East York             1.0        Sporting Goods Shop   
41      East Toronto             1.0           Greek Restaurant   
42      East Toronto             1.0       Fast Food Restaurant   
43      East Toronto             1.0                       Café   
45   Central Toronto             1.0                        Gym   
46   Central Toronto             1.0             Clothing Store   
47   Central Toronto             1.0               Dessert Shop   
49   Central Toronto             1.0                Coffee Shop   
51  Downtown Toronto             1.0                Coffee Shop   
52  Downtown Toronto             1.0                Coffee Shop   
53  Downtown Toronto             1.0                Coffee Shop   
54  Downtown Toronto             1.0                Coffee Shop   
55  Downtown Toronto             1.0                Coffee Shop   
56  Downtown Toronto             1.0               Cocktail Bar   
57  Downtown Toronto             1.0                Coffee Shop   
58  Downtown Toronto             1.0                Coffee Shop   
59  Downtown Toronto             1.0                Coffee Shop   
60  Downtown Toronto             1.0                Coffee Shop   
61  Downtown Toronto             1.0                Coffee Shop   
62        North York             1.0           Sushi Restaurant   
63   Central Toronto             1.0                     Garden   
64   Central Toronto             1.0                      Trail   
65   Central Toronto             1.0             Sandwich Place   
66  Downtown Toronto             1.0                       Café   
67  Downtown Toronto             1.0                       Café   
68  Downtown Toronto             1.0            Airport Service   
69  Downtown Toronto             1.0                Coffee Shop   
70  Downtown Toronto             1.0                Coffee Shop   
71        North York             1.0             Clothing Store   
73              York             1.0                      Field   
75  Downtown Toronto             1.0              Grocery Store   
77      West Toronto             1.0                        Bar   
78      West Toronto             1.0                       Café   
82      West Toronto             1.0            Thai Restaurant   
83      West Toronto             1.0                  Gift Shop   
84      West Toronto             1.0                Pizza Place   
85  Downtown Toronto             1.0                Coffee Shop   
86       Mississauga             1.0               Intersection   
88         Etobicoke             1.0                       Café   
91         Etobicoke             1.0                  Locksmith   
92         Etobicoke             1.0                        

### cluster 3

In [126]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0   Scarborough             2.0  Fast Food Restaurant    Dim Sum Restaurant   
80         York             2.0  Fast Food Restaurant        Discount Store   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0     Falafel Restaurant           Event Space  Ethiopian Restaurant   
80        Sandwich Place         Women's Store            Donut Shop   

   6th Most Common Venue 7th Most Common Venue        8th Most Common Venue  \
0    Empanada Restaurant     Electronics Store  Eastern European Restaurant   
80                 Diner   Distribution Center                      Dog Run   

   9th Most Common Venue 10th Most Common Venue  
0    Dumpling Restaurant              Drugstore  
80      Doner Restaurant              Drugstore

### cluster 4

In [127]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
2     Scarborough             3.0   Rental Car Location   
4     Scarborough             3.0   Fried Chicken Joint   
10    Scarborough             3.0     Indian Restaurant   
13    Scarborough             3.0              Pharmacy   
15    Scarborough             3.0    Chinese Restaurant   
19     North York             3.0                  Café   
30     North York             3.0         Grocery Store   
31     North York             3.0         Grocery Store   
32     North York             3.0         Grocery Store   
33     North York             3.0         Grocery Store   
35      East York             3.0           Pizza Place   
36      East York             3.0                  Park   
39      East York             3.0     Indian Restaurant   
76   West Toronto             3.0              Pharmacy   
81           York             3.0         Grocery Store   
87   East Toronto             3.0               Brewery   
89      Etobicoke             3.0           Pizza Place   
95      Etobicoke             3.0          Liquor Store   
96     North York             3.0   Empanada Restaurant   
99      Etobicoke             3.0           Pizza Place   
100     Etobicoke             3.0                  Park   
101     Etobicoke             3.0         Grocery Store   
102     Etobicoke             3.0   Rental Car Location   

         2nd Most Common Venue  3rd Most Common Venue 4th Most Common Venue  \
2               Breakfast Spot     Mexican Restaurant     Electronics Store   
4                  Gas Station                   Bank      Hakka Restaurant   
10                   Pet Store  Vietnamese Restaurant    Chinese Restaurant   
13                 Pizza Place                   Bank   Fried Chicken Joint   
15        Fast Food Restaurant            Coffee Shop              Pharmacy   
19                        Bank     Chinese Restaurant   Japanese Restaurant   
30                        Park     Athletics & Sports               Airport   
31                        Park     Athletics & Sports               Airport   
32                        Park     Athletics & Sports               Airport   
33                        Park     Athletics & Sports               Airport   
35        Fast Food Restaurant                   Café        Breakfast Spot   
36                Skating Rink                    Spa              Pharmacy   
39                         Gym             Restaurant           Pizza Place   
76                      Bakery            Pizza Place                   Gym   
81                 Pizza Place   Caribbean Restaurant         Women's Store   
87              Farmers Market          Auto Workshop         Burrito Place   
89                         Gym     Athletics & Sports              Pharmacy   
95                 Pizza Place                   Café        Cosmetics Shop   
96                 Pizza Place          Women's Store      Doner Restaurant   
99   Middle Eastern Restaurant         Discount Store          Intersection   
100                Pizza Place         Sandwich Place     Mobile Phone Shop   
101                 Beer Store    Japanese Restaurant        Discount Store   
102                  Drugstore          Women's Store      Doner Restaurant   

    5th Most Common Venue       6th Most Common Venue  \
2                    Bank                Intersection   
4    Caribbean Restaurant          Athletics & Sports   
10       Doner Restaurant          Dim Sum Restaurant   
13           Noodle House         Rental Car Location   
15          Shopping Mall                Burger Joint   
19     Dim Sum Restaurant              Discount Store   
30   Gym / Fitness Center  Construction & Landscaping   
31   Gym / Fitness Center  Construction & Landscaping   
32   Gym / Fitness Center  Construction & Landscaping   
33   Gym / Fitness Center  Construction & Landscaping   
35           Intersection          Athletics & Sports   
36          

### cluster 5

In [128]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
14       Scarborough             4.0                  Park   
20        North York             4.0             Cafeteria   
23        North York             4.0                  Park   
25        North York             4.0                  Park   
40         East York             4.0                  Park   
44   Central Toronto             4.0                  Park   
50  Downtown Toronto             4.0                  Park   
74              York             4.0                  Park   
79        North York             4.0                  Park   
90         Etobicoke             4.0                  Park   
98              York             4.0                  Park   

   2nd Most Common Venue       3rd Most Common Venue    4th Most Common Venue  \
14            Playground                  Donut Shop       Dim Sum Restaurant   
20                  Park        Colombian Restaurant  Comfort Food Restaurant   
23                  Bank                         Bar        Convenience Store   
25       Fireworks Store           Food & Drink Shop               Donut Shop   
40         Metro Station           Convenience Store               Donut Shop   
44           Swim School                    Bus Line         Doner Restaurant   
50            Playground                       Trail        Electronics Store   
74                  Pool               Women's Store                  Airport   
79                Bakery  Construction & Landscaping                Drugstore   
90                  Pool                       River                  Dog Run   
98     Convenience Store                   Drugstore                    Diner   

          5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
14                        Diner        Discount Store   Distribution Center   
20                  Event Space  Ethiopian Restaurant   Empanada Restaurant   
23                   Donut Shop        Discount Store   Distribution Center   
25                        Diner        Discount Store   Distribution Center   
40                        Diner        Discount Store   Distribution Center   
44                        Diner        Discount Store   Distribution Center   
50  Eastern European Restaurant   Dumpling Restaurant             Drugstore   
74               Farmers Market           Event Space  Ethiopian Restaurant   
79               Discount Store   Distribution Center               Dog Run   
90                 Dessert Shop    Dim Sum Restaurant                 Diner   
98               Discount Store   Distribution Center               Dog Run   

   8th Most Common Venue        9th Most Common Venue  \
14               Dog Run             Doner Restaurant   
20     Electronics Store  Eastern European Restaurant   
23               Dog Run             Doner Restaurant   
25               Dog Run             Doner Restaurant   
40               Dog Run             Doner Restaurant   
44               Dog Run                   Donut Shop   
50   Empanada Restaurant                   Donut Shop   
74   Empanada Restaurant            Electronics Store   
79      Doner Restaurant                   Donut Shop   
90        Discount Store          Distribution Center   
98      Doner Restaurant                   Donut Shop   

         10th Most Common Venue  
14                    Drugstore  
20          Dumpling Restaurant  
23                    Drugstore  
25                    Drugstore  
40                    Drugstore  
44           Falafel Restaurant  
50                Deli / Bodega  
74  Eastern European Restaurant  
79          Dumpling Restaurant  
90             Doner Restaurant  
98                Women's Store